kubetest2/kubetest2-gke/deployer/commandutils.go

```diff
@@ -0,0 +1,159 @@
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"k8s.io/test-infra/kubetest2/pkg/exec"
+)
+
+func (d *deployer) prepareGcpIfNeeded() error {
+	//TODO(RonWeber): boskos
+	if err := os.Setenv("CLOUDSDK_CORE_PRINT_UNHANDLED_TRACEBACKS", "1"); err != nil {
+		return fmt.Errorf("could not set CLOUDSDK_CORE_PRINT_UNHANDLED_TRACEBACKS=1: %v", err)
+	}
+	endpoint, err := d.endpoint()
+	if err != nil {
+		return err
+	}
+	if err := os.Setenv("CLOUDSDK_API_ENDPOINT_OVERRIDES_CONTAINER", endpoint); err != nil {
+		return err
+	}
+
+	if err := runWithOutput(exec.Command("gcloud", "config", "set", "project", d.project)); err != nil {
+		return fmt.Errorf("Failed to set project %s : err %v", d.project, err)
+	}
+
+	// gcloud creds may have changed
+	if err := activateServiceAccount(d.gcpServiceAccount); err != nil {
+		return err
+	}
+
+	// Ensure ssh keys exist
+	log.Print("Checking existing of GCP ssh keys...")
+	k := filepath.Join(home(".ssh"), "google_compute_engine")
+	if _, err := os.Stat(k); err != nil {
+		return err
+	}
+	pk := k + ".pub"
+	if _, err := os.Stat(pk); err != nil {
+		return err
+	}
+
+	// Install custom gcloud version if necessary
+	if d.gcpCloudSDK != "" {
+		for i := 0; i < 3; i++ {
+			if err := runWithOutput(exec.Command("gsutil", "-mq", "cp", "-r", d.gcpCloudSDK, home())); err == nil {
+				break // Success!
+			}
+			time.Sleep(1 << uint(i) * time.Second)
+		}
+		for _, f := range []string{home(".gsutil"), home("repo"), home("cloudsdk")} {
+			if _, err := os.Stat(f); err == nil || !os.IsNotExist(err) {
+				if err = os.RemoveAll(f); err != nil {
+					return err
+				}
+			}
+		}
+
+		install := home("repo", "google-cloud-sdk.tar.gz")
+		if strings.HasSuffix(d.gcpCloudSDK, ".tar.gz") {
+			install = home(filepath.Base(d.gcpCloudSDK))
+		} else {
+			if err := os.Rename(home(filepath.Base(d.gcpCloudSDK)), home("repo")); err != nil {
+				return err
+			}
+
+			// Controls which gcloud components to install.
+			pop, err := pushEnv("CLOUDSDK_COMPONENT_MANAGER_SNAPSHOT_URL", "file://"+home("repo", "components-2.json"))
+			if err != nil {
+				return err
+			}
+			defer pop()
+		}
+
+		if err := installGcloud(install, home("cloudsdk")); err != nil {
+			return err
+		}
+		// gcloud creds may have changed
+		if err := activateServiceAccount(d.gcpServiceAccount); err != nil {
+			return err
+		}
+	}
+
+	//TODO(RonWeber): kubemark
+	return nil
+}
+
+// Activate service account if set or do nothing.
+func activateServiceAccount(path string) error {
+	if path == "" {
+		return nil
+	}
+	return runWithOutput(exec.Command("gcloud", "auth", "activate-service-account", "--key-file="+path))
+}
+
+// home returns $HOME/part/part/part
+func home(parts ...string) string {
+	p := []string{os.Getenv("HOME")}
+	for _, a := range parts {
+		p = append(p, a)
+	}
+	return filepath.Join(p...)
+}
+
+// pushEnv pushes env=value and return a function that resets env
+func pushEnv(env, value string) (func() error, error) {
```

Comment:
we should probably refactor this out to just pass it to the target process, popping env is a bit of an antipattern

Reply:
Done.